### ISSUE TITLE
[Feature]: Retry on failure functionality

### ISSUE THREAD

**Issue by neubig** [OpenHands/OpenHands#2221]
**What problem or use case are you trying to solve?**

Sometimes models fail to do their job correctly, and we would benefit from starting all over from the beginning. There are a few examples of this in the agent literature:
- Aider recently introduced a [harness](https://github.com/paul-gauthier/aider-swe-bench) for testing SWE-bench that allows for retries when tests and linting don't pass on swe-bench
- @Jiayi-Pan has work on [Evaluation and Refinement for web agents](https://arxiv.org/pdf/2404.06474) that use a reward model to judge when a web task has failed, a reset mechanism to return to the beginning, and a method for improving the prompts based on reflexion.
- @niansong1996 has a method [LEVER](https://github.com/niansong1996/lever) that uses a learned verifier to rerank code generation execution results.

**Describe the UX of the solution you'd like**

Ideally, this would be something that could be implemented in a general way, so that we could implement different strategies with a shared interface. For instance:

```
class ResetStrategy:

   @abstractmethod
   def initialize_state():
       """Take note of the initial state that should be reset too."""
       ...

   @abstractmethod
   def verify(...):
      """This verifies whether the agent has reached a failure state."""
      ...
   
   @abstractmethod
   def reset(...):
      """This performs some sort of reset."""
      ...

   @abstractmethod
   def message_on_reset(...):
      """This creates a message to the agent upon reset (e.g. a task with a prompt based on reflexion)."""
      ...   
```

Then, when using OpenDevin, we could choose an option that says "retry N times when you get stuck", and select the strategy that is used to do so.

**Do you have thoughts on the technical implementation?**

The actual reset strategies would vary based on the task. For instance:

- AiderResetStrategy ([code reference](https://github.com/paul-gauthier/aider-swe-bench/blob/1ae32650e44707c1aa67bbe0a7bf090edd1358c6/harness.py#L229-L339)):
  - `initialize`: save the current git commit of the repository `commit_id`
  - `verify`: tests+linting pass
  - `reset`: `git checkout commit_id`
  - `message_on_failure`: no-op
- EvalRefineResetStrategy ([code reference](https://github.com/Berkeley-NLP/Agent-Eval-Refine)):
  - `initialize`: save the current web page `initial_page`
  - `verify`: the reward model is positive
  - `reset`: `goto(initial_page)`
  - `message_on_failure`: reflexion prompt

This could either be integrated into OpenDevin, allowing for retries in the main app as well

**Additional context:**

* [Aider superissue](https://github.com/OpenDevin/OpenDevin/issues/120)
* Thanks to @xingyaoww and @frankxu2004 for offline discussion

### If you find this feature request or enhancement useful, make sure to add a 👍 to the issue

**Comment by Jiayi-Pan** [comment]
Thanks for creating the issue! 
Although I don’t have much spare bandwidth recently, I am definitely interested in bringing EvalRefineResetStrategy and the retry functionality into OpenDevin. I will keep an eye on this PR and contribute once I have the time.

**Comment by Vaishakh-SM** [comment]
Hi!
Is anyone working on this?

**Comment by neubig** [comment]
Hey @Vaishakh-SM , I think nobody is working on this, but @xingyaoww was thinking about adding multiple runs to evaluation. I think that would be a parallel effort though, because it would involve running multiple times and picking the best one, as opposed to restarting when the first try didn't work.

If you'd be interested in taking a look it'd be welcome!

**Comment by Vaishakh-SM** [comment]
This seems like an interesting problem!

I'll take a look and get back to this sometime this week.

**Comment by mamoodi** [comment]
@neubig this is a really old issue. Just want to make sure, we haven't implemented this yet, right?

**Comment by neubig** [comment]
Yep, @xingyaoww is working on a critic that could help implement this.

**Comment by manzke** [comment]
Using OpenHands now for a whole project, I just can say a retry in general would be nice. I've seen several failures, when things could not be replaced or far more often, if they could not be generated because of rate limits of the models. The bigger your codebase gets, the more tokens are used and the more often you hit the rate limits.
This led to several cases, where files had been deleted, but couldn't be generated again. Happy to share more insights.

**Comment by neubig** [comment]
We're still working on this

**Comment by mamoodi** [comment]
Is this still being actively worked on?

**Comment by neubig** [comment]
Yes, it's still part of the critic that @xingyaoww is working on

**Comment by mamoodi** [comment]
@neubig and @enyst would this fall under agent-sdk at all?

**Comment by jpelletier1** [comment]
@xingyaoww Checking to see if this is still part of the Critic model?

**Comment by xingyaoww** [comment]
@jpelletier1 Yep. We just need to figure out the correct user experience for it. Initial critic support is ready for review now:

SDK: https://github.com/OpenHands/software-agent-sdk/pull/1269
Docs: https://github.com/OpenHands/docs/pull/263
CLI: https://github.com/OpenHands/OpenHands-CLI/pull/360

**Comment by mkdev5** [comment]
@jpelletier1 Can I work on this?

**Comment by jpelletier1** [comment]
@MkDev11 I am not the subject matter expert on this one, unfortunately. It looks like it's in progress (maybe with @xingyaoww) so suggest taking guidance from him before proceeding.

**Comment by mkdev5** [comment]
@xingyaoww Please let me know if I can work on this. Thanks.

**Comment by xingyaoww** [comment]
Hi @MkDev11 , we've recently just shipped critic - https://docs.openhands.dev/openhands/usage/cli/critic

and the example for iterative refinement is also ready for review soon here: https://github.com/OpenHands/software-agent-sdk/pull/1879

So i'd consider this is already implemented in the core agent SDK :(

But feel free to contribute to support it in the application if you'd like!

**Comment by xingyaoww** [comment]
And as a follow-up to the critic model, we did need to implement something in the SDK similar to what @neubig originally proposed on this thread: https://github.com/OpenHands/software-agent-sdk/pull/1706

Feel free to check the above PR and contribution a potential abstraction to the SDK as well :)

**Comment by neubig** [comment]
I think that maybe we should close this. It's not super-obvious how we can implement it in the application, so we can discuss more and then open a well-specified issue when we know what to do.

**Comment by enyst** [comment]
This is lovely though! An issue from June 3rd 2024! ❤️ 

Not forgotten, it changed but not really stale, worth trying it over, thank you!
